database/src/service/write.rs

```diff
@@ -0,0 +1,210 @@
+    type Future = ResponseRecv;
+
+    #[inline]
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        todo!()
+    }
+
+    #[inline]
+    fn call(&mut self, _req: WriteRequest) -> Self::Future {
+        todo!()
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- DatabaseWriter
+/// Database writer thread.
+///
+/// Each reader thread is spawned with access to this struct (self).
+///
+pub(super) struct DatabaseWriter {
+    /// Receiver side of the database request channel.
+    ///
+    /// Any caller can send some requests to this channel.
+    /// They send them alongside another `Response` channel,
+    /// which we will eventually send to.
+    receiver: crossbeam::channel::Receiver<(WriteRequest, ResponseSend)>,
+
+    /// TODO: either `Arc` or `&'static` after `Box::leak`
+    /// Access to the database.
+    db: &'static ConcreteEnv,
+}
+
+impl DatabaseWriter {
+    /// Initialize the `DatabaseWriter` thread-pool.
+    ///
+    /// This spawns `N` amount of `DatabaseWriter`'s
+    /// attached to `db` and returns a handle to the pool.
+    ///
+    /// The reason this isn't a single thread is to allow a little
+    /// more work to be done during the leeway in-between DB operations, e.g:
+    ///
+    /// ```text
+    ///  Request1
+    ///     |                                      Request2
+    ///     v                                         |
+    ///  Writer1                                      v
+    ///     |                                      Writer2
+    /// doing work                                    |
+    ///     |                               waiting on other writer
+    /// sending response  --|                         |
+    ///     |               |- leeway            doing work
+    ///     v             --|                         |
+    ///    done                                       |
+    ///                                        sending response
+    ///                                               |
+    ///                                               v
+    ///                                              done
+    /// ```
+    ///
+    /// During `leeway`, `Writer1` is:
+    /// - busy preparing the message
+    /// - creating the response channel
+    /// - sending it back to the channel
+    /// - ...etc
+    ///
+    /// During this time, it would be wasteful if `Request2`'s was
+    /// just being waited on, so instead, `Writer2` handles that
+    /// work while `Writer1` is in-between tasks.
+    ///
+    /// This leeway is pretty tiny (doesn't take long to allocate a channel
+    /// and send a response) so there's only 2 Writers for now (needs testing).
+    ///
+    /// The database backends themselves will hang on write transactions if
+    /// there are other existing ones, so we ourselves don't need locks.
```

Comment:
Although this is a good idea, only one place will write to the DB and as the requestor *needs* to await the response (otherwise the operation could have failed) you should never have 2 write requests in the channel.

So I think there should only be 1 write thread.